src/combinatorics/simple_graph/metric.lean

```diff
@@ -0,0 +1,102 @@
+/-
+Copyright (c) 2022 Kyle Miller. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Kyle Miller, Vincent Beffara
+-/
+import combinatorics.simple_graph.connectivity
+import data.nat.lattice
+
+/-!
+# Graph metric
+
+This module defines the `simple_graph.dist` function, which takes
+pairs of vertices to the length of the shortest walk between them.
+
+## Main definitions
+
+- `simple_graph.dist` is the graph metric.
+
+## Todo
+
+- Provide an additional computable version of `simple_graph.dist`
+  for when `G` is connected.
+
+## Tags
+
+graph metric
+
+-/
+
+namespace simple_graph
+variables {V : Type*} (G : simple_graph V)
+
+/-! ## Metric -/
+
+/-- The distance between two vertices is the length of the shortest walk between them.
+If no such walk exists, this uses the junk value of `0`. -/
+noncomputable
+def dist (u v : V) : ℕ := Inf (set.range (walk.length : G.walk u v → ℕ))
+
+variables {G}
+
+protected
+lemma reachable.exists_walk_of_dist {u v : V} (hr : G.reachable u v) :
+  ∃ (p : G.walk u v), p.length = G.dist u v :=
+nat.Inf_mem (set.range_nonempty_iff_nonempty.mpr hr)
+
+protected
+lemma connected.exists_walk_of_dist (hconn : G.connected) (u v : V) :
+  ∃ (p : G.walk u v), p.length = G.dist u v :=
+(hconn u v).exists_walk_of_dist
+
+lemma dist_le {u v : V} (p : G.walk u v) : G.dist u v ≤ p.length := nat.Inf_le ⟨p, rfl⟩
+
+@[simp] lemma dist_self {v : V} : dist G v v = 0 :=
+le_antisymm (dist_le (walk.nil : walk G v v)) (zero_le _)
+
+@[simp]
+lemma dist_eq_zero_iff_eq_or_not_reachable {u v : V} : G.dist u v = 0 ↔ u = v ∨ ¬ G.reachable u v :=
+by simp [dist, nat.Inf_eq_zero, reachable]
+
+protected
+lemma reachable.dist_eq_zero_iff {u v : V} (hr : G.reachable u v) :
+  G.dist u v = 0 ↔ u = v := by simp [hr]
+
+protected
+lemma connected.dist_eq_zero_iff (hconn : G.connected) {u v : V} :
+  G.dist u v = 0 ↔ u = v := by simp [hconn u v]
+
+lemma dist_eq_zero_of_not_reachable {u v : V} (h : ¬ G.reachable u v) : G.dist u v = 0 :=
+by simp [h]
+
+lemma nonempty_of_pos_dist {u v : V} (h : 0 < G.dist u v) :
+  (set.univ : set (G.walk u v)).nonempty :=
+by simpa [set.range_nonempty_iff_nonempty, set.nonempty_iff_univ_nonempty]
+     using nat.nonempty_of_pos_Inf h
+
+protected
+lemma connected.dist_triangle (hconn : G.connected) {u v w : V} :
+  G.dist u w ≤ dist G u v + dist G v w :=
+begin
+  obtain ⟨p, hp⟩ := hconn.exists_walk_of_dist u v,
+  obtain ⟨q, hq⟩ := hconn.exists_walk_of_dist v w,
+  rw [← hp, ← hq, ← walk.length_append],
+  apply dist_le,
+end
+
+lemma dist_comm_le {u v : V} (h : G.reachable u v) : G.dist u v ≤ G.dist v u :=
```

Comment:
Do you want this to be private? Or maybe proved inside `dist_comm`, since it's very much weaker than that

Reply:
That would make sense, yes.

Reply:
I changed it to `dist_comm_aux` and made it `private`. (I made the edit directly in the github.dev editor, so hopefully without breaking it!)